### PR TITLE
[libunwind] Replace process_vm_readv with SYS_rt_sigprocmask

### PR DESCRIPTION
**Pull request by ajordanr-google**
process_vm_readv is generally considered dangerous from a syscall perspective, and is frequently blanket banned in seccomp filters such as those in Chromium and ChromiumOS. We can get the same behaviour during the invalid PC address case with the raw SYS_rt_sigprocmask syscall.

Testing to ensure that process_vm_readv does not appear, I ran the output of check-unwind on an ARM64 device under strace. Previously, bad_unwind_info in particular would use process_vm_readv, but with this commit, it now no longer uses it:

```
strace test/Output/bad_unwind_info.pass.cpp.dir/t.tmp.exe \
  |& grep process_vm_readv
```

The libunwind unittests were also tested on ARM64 ChromeOS (Gentoo Linux) devices.

### CODE REVIEW

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2713]: Consider SYS_rt_sigprocmask like absl/debugging/internal/address_is_readable.cc

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Fixed.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [reply libunwind/src/UnwindCursor.hpp:2713]: A pipe has the downside of requiring a new file descriptor. The absl migration away from a pipe was partly motivated by performance and partly by avoiding a new file descriptor.

Other OSes use their own implementations if needed, so we do not need to worry about the portability. 

Sorry that I am not familiar with SYS_rt_procmask more, so I probably cannot answer some questions, but I suppose that these properties can be straightforwardly tested using a local small test.

Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp:2713]: Gotcha, that all makes sense. Thanks!

---

Playing around with `SYS_rt_procmask` on Linux x86_64, There's apparently a point in which we can still read a 
given address but `SYS_rt_procmask` sets `errno = 14` (Bad Address)

Looping over the raw syscall with:

```
syscall(SYS_rt_sigprocmask, /* how= */ -1, Addr, Addr, 8);
````

until it breaks gives

```
Addr: 0x56322ab46ff8; Offset: 59704; errno: 22; Value: 0x0
Addr: 0x56322ab46ff9; Offset: 59705; errno: 14; Value: 0x0
Addr: 0x56322ab46ffa; Offset: 59706; errno: 14; Value: 0x0
Addr: 0x56322ab46ffb; Offset: 59707; errno: 14; Value: 0x0
Addr: 0x56322ab46ffc; Offset: 59708; errno: 14; Value: 0x0
Addr: 0x56322ab46ffd; Offset: 59709; errno: 14; Value: 0x0
Addr: 0x56322ab46ffe; Offset: 59710; errno: 14; Value: 0x0
Addr: 0x56322ab46fff; Offset: 59711; errno: 14; Value: 0x0
fish: Job 1, './main' terminated by signal SIGSEGV (Address boundary error)
```

Seems there's about a 7 to 8 byte boundary where it fails but we can still read, at least on x86_64. So the 8 byte alignment on AARCH64 as mentioned in address_is_readable.cc seems to make sense, though we'd probably want to still check both the beginning and end of the `instructions` range. I think that should be sufficient.

File: libunwind/test/bad_unwind_info.pass.cpp
Comment (ajordanr-google) [inline libunwind/test/bad_unwind_info.pass.cpp:29]: Note: changed this for an 8 byte alignment check, because `0x4 & 0x7` is `0x0`, whereas `0xC & 0x7` is `0x8`.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2762]: It is unfortunate that we have to call `isReadableAddr` twice.

`linux/kernel/signal.c` `rt_sigprocmask` actually supports an unaligned address. We can remove the alignment code `const auto alignedAddr = addr & ~pint_t{7};` and use one `isReadableAddr(pc)`

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2820]: Once call suffices as one rt_sigprocmask call checks 8 bytes.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2989]: `sizeof(unsigned long)` for 32-bit ELF targets. We should use 8 (tested on i386).

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2972]: The code is significantly different so the license (Apache2) comment should not be needed.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2995]: rt_sigprocmask is a Linux-specific syscall,and the guard macro `_LIBUNWIND_CHECK_LINUX_SIGRETURN` is about Linux, so the "Linux specific" comment is unneeded

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2991]: `oldsigset` should be nullptr, otherwise also checks that `sigsetAddr` is writable

File: libunwind/src/UnwindCursor.hpp
Comment (arichardson) [inline libunwind/src/UnwindCursor.hpp:2993]: I would add a comment that this code relies on the Linux kernel checking copy_from_user success before checking the how argument is valid. I think this is somewhat fragile but I guess the kernel can't change this behaviour anymore without breaking userspace.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp:2989]: Thanks! If you have a better idea on how to make this more robust, I'd love to know. For now I've hardcoded it to 8 then.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp:2972]: Done.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Done

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp:2989]: Thanks, done.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Removed, thanks.

File: libunwind/src/UnwindCursor.hpp
Comment (arichardson) [reply libunwind/src/UnwindCursor.hpp:2989]: It is important that we get this right since otherwise the syscall returns EINVAL without trying to access the user memory: https://elixir.bootlin.com/linux/v6.6.7/source/kernel/signal.c#L3184

File: libunwind/src/UnwindCursor.hpp
Comment (arichardson) [reply libunwind/src/UnwindCursor.hpp:2991]: The check for valid `how` happens first (https://elixir.bootlin.com/linux/v6.6.7/source/kernel/signal.c#L3184) so we get EINVAL before this is checked. However, I agree there is no need to pass this argument.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Given Paul's offline comments, I've removed the alignment as it seems to be a bug in QEMU not Linux.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Done

File: libunwind/src/UnwindCursor.hpp
Comment (arichardson) [inline libunwind/src/UnwindCursor.hpp:2820]: We might get a false-negative `return false` here if the instruction happens to be the last four bytes before the end of the page boundary but I don't think it it matters.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Looking at linux-6.6.7 source code, we have some minor variety on `sigset_t` size...

TL;DR: They are almost all `_NSIG` bits long, which is 64 bits or, in the case of MIPS, maybe 128? There are two exceptions to this, where they are not a struct of size `_NSIG/_NSIG_BPW * sizeof(unsigned long)` but rather just `unsigned long`. I assume `_NSIG_BPW == sizeof(long) == sizeof(unsigned long)`, which seems reasonable given the comments (and not `_NSIG_BPW == sizeof(int)`).

---

`sigset_t` is almost always defined as:

```
typedef struct {
        unsigned long sig[_NSIG_WORDS];
} sigset_t;
```

Where:

```
arch/parisc/include/asm/signal.h-7-#define _NSIG                64
arch/parisc/include/asm/signal.h-8-/* bits-per-word, where word apparently means 'long' not 'int' */
arch/parisc/include/asm/signal.h-9-#define _NSIG_BPW    BITS_PER_LONG
arch/parisc/include/asm/signal.h:10:#define _NSIG_WORDS (_NSIG / _NSIG_BPW)

// or

arch/xtensa/include/uapi/asm/signal.h-18-#define _NSIG          64
arch/xtensa/include/uapi/asm/signal.h-19-#define _NSIG_BPW      32
arch/xtensa/include/uapi/asm/signal.h:20:#define _NSIG_WORDS    (_NSIG / _NSIG_BPW)

// or

arch/ia64/include/asm/signal.h-15-#define _NSIG         64
arch/ia64/include/asm/signal.h-16-#define _NSIG_BPW     64
arch/ia64/include/asm/signal.h:17:#define _NSIG_WORDS   (_NSIG / _NSIG_BPW)

// or

arch/alpha/include/asm/signal.h-10-#define _NSIG                64
arch/alpha/include/asm/signal.h-11-#define _NSIG_BPW    64
arch/alpha/include/asm/signal.h:12:#define _NSIG_WORDS  (_NSIG / _NSIG_BPW)

// or

arch/arm/include/asm/signal.h-10-#define _NSIG          64
arch/arm/include/asm/signal.h-11-#define _NSIG_BPW      32
arch/arm/include/asm/signal.h:12:#define _NSIG_WORDS    (_NSIG / _NSIG_BPW)

// or

arch/s390/include/asm/signal.h-12-/* Most things should be clean enough to redefine this at will, if care
arch/s390/include/asm/signal.h-13-   is taken to make libc match.  */
arch/s390/include/asm/signal.h-14-#include <asm/sigcontext.h>
arch/s390/include/asm/signal.h-15-#define _NSIG           _SIGCONTEXT_NSIG
arch/s390/include/asm/signal.h-16-#define _NSIG_BPW       _SIGCONTEXT_NSIG_BPW
arch/s390/include/asm/signal.h:17:#define _NSIG_WORDS     _SIGCONTEXT_NSIG_WORDS
arch/s390/include/uapi/asm/sigcontext.h:30:/* Has to be at least _NSIG_WORDS from asm/signal.h */
arch/s390/include/uapi/asm/sigcontext.h-31-#define _SIGCONTEXT_NSIG     64
arch/s390/include/uapi/asm/sigcontext.h-32-#define _SIGCONTEXT_NSIG_BPW 64

// or

arch/x86/include/asm/signal.h-11-#define _NSIG          64
arch/x86/include/asm/signal.h-12-
arch/x86/include/asm/signal.h-13-#ifdef __i386__
arch/x86/include/asm/signal.h-14-# define _NSIG_BPW     32
arch/x86/include/asm/signal.h-15-#else
arch/x86/include/asm/signal.h-16-# define _NSIG_BPW     64
arch/x86/include/asm/signal.h-17-#endif
arch/x86/include/asm/signal.h-18-
arch/x86/include/asm/signal.h:19:#define _NSIG_WORDS    (_NSIG / _NSIG_BPW)

// or

arch/m68k/include/asm/signal.h-10-#define _NSIG         64
arch/m68k/include/asm/signal.h-11-#define _NSIG_BPW     32
arch/m68k/include/asm/signal.h:12:#define _NSIG_WORDS   (_NSIG / _NSIG_BPW)

// or (quite weirdly)

arch/mips/include/uapi/asm/signal.h-15-#define _NSIG            128
arch/mips/include/uapi/asm/signal.h-16-#define _NSIG_BPW        (sizeof(unsigned long) * 8)
arch/mips/include/uapi/asm/signal.h:17:#define _NSIG_WORDS      (_NSIG / _NSIG_BPW)

// or

arch/sparc/include/uapi/asm/signal.h-81-#define __NEW_NSIG      64
arch/sparc/include/uapi/asm/signal.h-82-#ifdef __arch64__
arch/sparc/include/uapi/asm/signal.h-83-#define _NSIG_BPW       64
arch/sparc/include/uapi/asm/signal.h-84-#else
arch/sparc/include/uapi/asm/signal.h-85-#define _NSIG_BPW       32
arch/sparc/include/uapi/asm/signal.h-86-#endif
arch/sparc/include/uapi/asm/signal.h:87:#define _NSIG_WORDS     (__NEW_NSIG / _NSIG_BPW)

// or

arch/powerpc/include/uapi/asm/signal.h-7-#define _NSIG          64
arch/powerpc/include/uapi/asm/signal.h-8-#ifdef __powerpc64__
arch/powerpc/include/uapi/asm/signal.h-9-#define _NSIG_BPW      64
arch/powerpc/include/uapi/asm/signal.h-10-#else
arch/powerpc/include/uapi/asm/signal.h-11-#define _NSIG_BPW     32
arch/powerpc/include/uapi/asm/signal.h-12-#endif
arch/powerpc/include/uapi/asm/signal.h:13:#define _NSIG_WORDS   (_NSIG / _NSIG_BPW)

// or (finally)

include/uapi/asm-generic/signal.h-7-#define _NSIG               64
include/uapi/asm-generic/signal.h-8-#define _NSIG_BPW   __BITS_PER_LONG
include/uapi/asm-generic/signal.h:9:#define _NSIG_WORDS (_NSIG / _NSIG_BPW)
```

Here are the two exceptions to that (and their line numbers):

```
arch/x86/include/uapi/asm/signal.h:17:typedef unsigned long sigset_t;
arch/alpha/include/uapi/asm/signal.h:14:typedef unsigned long sigset_t;
```

There were also various `old_sigset_t` and `__kernel_sigset_t`, which were defined like:

```
arch/ia64/include/asm/signal.h:24:typedef unsigned long old_sigset_t;
arch/alpha/include/asm/signal.h:14:typedef unsigned long old_sigset_t;          /* at least 32 bits */
arch/m68k/include/asm/signal.h:14:typedef unsigned long old_sigset_t;           /* at least 32 bits */
arch/s390/include/asm/signal.h:19:typedef unsigned long old_sigset_t;             /* at least 32 bits */
arch/x86/include/asm/signal.h:21:typedef unsigned long old_sigset_t;            /* at least 32 bits */
arch/parisc/include/asm/signal.h:17:typedef unsigned long old_sigset_t;         /* at least 32 bits */
arch/mips/include/uapi/asm/signal.h:23:typedef unsigned long old_sigset_t;              /* at least 32 bits */
arch/powerpc/include/uapi/asm/signal.h:15:typedef unsigned long old_sigset_t;           /* at least 32 bits */
```

and

```
arch/ia64/include/uapi/asm/posix_types.h:5:typedef unsigned long        __kernel_sigset_t;      /* at least 32 bits */
arch/s390/include/uapi/asm/posix_types.h:46:typedef unsigned long   __kernel_sigset_t;      /* at least 32 bits */
arch/alpha/include/uapi/asm/posix_types.h:14:typedef unsigned long      __kernel_sigset_t;      /* at least 32 bits */
```

Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: How about `(NSIG-1)/8`? https://www.gnu.org/software/libc/manual/html_node/Standard-Signals.html. Just let libc handle it. It should work cleanly for everything that's not MIPS. I don't have a MIPS system to test if this works for it.

Comment (MaskRay) [reply libunwind/src/UnwindCursor.hpp]: I think we should not worry about mips. mips uses are being phased out, and the use cases (embedded?) seems to not care about this involved sigreturn trampoline anyway...

File: libunwind/test/bad_unwind_info.pass.cpp
Comment (ajordanr-google) [reply libunwind/test/bad_unwind_info.pass.cpp]: Undid this with the removal of the alignment code.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp:2820]: I think that's probably fine, as long as we aren't hitting a SIGSEGV.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Gotcha. I ended up using `NSIG/8` since it rounds down anyways, and I checked the glibc code and it has 128 instead of 127, and I think we would actually want 128/8 instead if I'm reading the mips case right. It's not that important if we don't care about mips.

File: libunwind/src/UnwindCursor.hpp
Comment (arichardson) [inline libunwind/src/UnwindCursor.hpp:2996]: Do we need to save+restore errno? Or is libunwind free to modify errno?

**Review by arichardson — APPROVED** [review]
Just one question otherwise this looks good to me.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2991]: "EFAULT is not guaranteed by the POSIX standard." should be removed since the whole function is Linux specific

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [reply libunwind/src/UnwindCursor.hpp:2989]: Feel free to use `maybe_unused`. libunwind is built with C++17 now.

File: libunwind/src/UnwindCursor.hpp
Comment (MaskRay) [inline libunwind/src/UnwindCursor.hpp:2973]: I think it is fine to just say without more quote about the license.

We use SYS_rt_sigprocmask, inspired by Abseil's AddressIsReadable.

**Review by MaskRay — APPROVED** [review]

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Done

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: Done.

File: libunwind/src/UnwindCursor.hpp
Comment (ajordanr-google) [reply libunwind/src/UnwindCursor.hpp]: We should probably save and restore, as we do this above in `getInfoFromTBTable`. Good point. Done.

**Comment by ajordanr-google** [comment]
I don't have Commit Access (I so rarely actually commit to upstream), and I don't think I have yet the "track record of submitting high quality patches" :)

Can someone else merge this once the checks are green?

**Comment by ajordanr-google** [comment]
Hmm. It looks like `[[maybe_unused]]` is still disallowed...

**Comment by MaskRay** [comment]
> Hmm. It looks like `[[maybe_unused]]` is still disallowed...

`[[maybe_unused]]` should be allowed when the pull request includes commit 47413bb2760e63a3302871ea770d6c0f5a742036 (C++17 for libunwind).

**Comment by ajordanr-google** [comment]
Rebased to include 47413bb2760e63a3302871ea770d6c0f5a742036 (a merge commit seemed silly?). No other changes have been made.

**Comment by ajordanr-google** [comment]
All CI runs are failing on unrelated tests... can we re-run the CI again in a week or so?

**Pull request merged**
